### ISSUE TITLE
Fail to apply equality to side-condition

### ISSUE THREAD

**Issue by ehildenb** [kframework/kore#2849]
I have a side-condition that says something like `{ false #Equals PRED } #And { true #Equals (PRED impliesBool ELSE) }`, and I would expect it to replace the second `PRED` with `false` and then continue simplifying, but it does not.

K version:

```
% kompile --version
K version:    v5.1.188-4-g6f2b343a8f
Build date:   Tue Sep 21 16:51:08 UTC 2021
```

Kore version:

```
98e241ede
```

Input K file:

```
module TEST
  imports INT
  imports MAP
  imports BOOL

  configuration <k> $PGM:Step </k>
                <mem> .Map </mem>
                <amount> 0 </amount>
                <value> 0 </value>
                <key> 0:KItem </key>

  syntax Step ::= "a" | "b" | "c" | "z"

  rule <k> a => b ... </k> <amount> A </amount> requires         A ==Int 0
  rule <k> a => z ... </k> <amount> A </amount> requires notBool A ==Int 0

  rule <k> b => c ... </k> <mem> AS </mem> <key> K </key> requires notBool K in_keys(AS)
  rule <k> b => z ... </k> <mem> AS </mem> <key> K </key> requires         K in_keys(AS)
    
  syntax Int ::= #lookup ( Map , K ) [function, functional, no-evaluators]
  rule I ==Int #lookup(M, K) => ( ((notBool #isPresent(M, K)) impliesBool I ==Int 0           )
                          andBool (         #isPresent(M, K)  impliesBool I ==Int {M[K]}:>Int )
                                )
    [simplification]

  syntax Bool ::= #isPresent ( Map , K )
  rule #isPresent(M, K) => K in_keys(M) andBool isInt(M[K]) [macro]

endmodule

module TEST-SPEC
  imports TEST

  claim <k> a => z ... </k>
        <mem> Allownces </mem>
        <amount> Amount </amount>
        <value> Value </value>
        <key> K  </key>
    requires Allownce ==Int #lookup(Allowances, K)
     andBool notBool ( Amount ==Int 0
               andBool (Allowance >Int 0 impliesBool Value ==Int 0)
                     )
endmodule
```

Call commands:

```
kompile --backend haskell test.k --main-module TEST
kprovex test.k --spec-module TEST-SPEC
```

And it gives counter-example:

```
  <generatedTop>
    <k>
      c ~> _DotVar1 ~> .
    </k>
    <mem>
      Allownces
    </mem>
    <amount>
      0
    </amount>
    <value>
      Value
    </value>
    <key>
      K
    </key>
  </generatedTop>
#And
  {
    false
  #Equals
    Allowance >Int 0 impliesBool Value ==Int 0
  }
#And
  {
    false
  #Equals
    K in_keys ( Allownces )
  }
#And
  {
    true
  #Equals
    ( K in_keys ( Allowances ) andBool isInt ( Allowances [ K ] ~> . ) ) impliesBool Allownce ==Int project:Int ( Allowances [ K ] ~> . )
  }
#And
  {
    true
  #Equals
    notBool ( K in_keys ( Allowances ) andBool isInt ( Allowances [ K ] ~> . ) ) impliesBool Allownce ==Int 0
  }
```

In particular, we see `{ false #Equals K in_keys(Allowances) }`, and then several other side-conditions where substituting in `false` for that same predicate `K in_keys(Allowances)` would make it able to simplify further.

This is blocking Michelson K update.

Bug report attached.
[test.tar.gz](https://github.com/kframework/kore/files/7205255/test.tar.gz)

**Comment by ana-pantilie** [comment]
I'm pretty sure this is the same issue as https://github.com/kframework/kore/issues/2785, which has a fix ready but we need to get a performance update in first.

**Comment by ehildenb** [comment]
Note that adding rules like `rule A andBool _ => false requires notBool A [simplification]` does not solve this.

**Comment by ana-pantilie** [comment]
This should be fixed on ed422630f9383b2774f1ec8ea04eee86ea0afe1e. Please let me know if you still encounter issues.
